Read Short as u16, SignedShort as i16

weiji14 opened the pull request:
Previously, Short values were read incorrectly as `u32` instead of `u16`, and SignedShort values were read incorrectly as `i32` instead of `i16`.

This PR does the following:
- Add `TiffFormatError::ShortExpected` enum variant
- Correctly cast `Type::SHORT` as u16 values, and support casting Short to other unsigned types via the `into_u32_vec` and `into_u64_vec` methods.
- Correctly cast `Type::SSHORT` as i16 values

Fixes #204

weiji14 committed:
Add TiffFormatError::ShortExpected enum variant

For values expected to be u16.

weiji14 committed:
Cast Short to u16 instead of u32

Adapted from https://github.com/image-rs/image-tiff/commit/84ff6795e0400305ce86478d5b200b2350457479

weiji14 committed:
Cast SignedShort to i16 instead of i32

fintelia commented:
The hard part is still [this concern](https://github.com/image-rs/image-tiff/issues/204#issuecomment-1566616975):

> This makes sense to fix, though the annoying part is figuring out whether anywhere else in the code is relying on the current behavior. Or whether it is likely to break any downstream users and thus might warrant additional caution

weiji14 commented:
> The hard part is still [this concern](https://github.com/image-rs/image-tiff/issues/204#issuecomment-1566616975):
> 
> > This makes sense to fix, though the annoying part is figuring out whether anywhere else in the code is relying on the current behavior. Or whether it is likely to break any downstream users and thus might warrant additional caution

A `tiff::decoder::ifd::Value` would still need to be cast into a primitive type using `into_u16`, `into_u32` or some other method no? My impression was that the change here would mean a user doing `Value.into_u16` would save some memory by not going through `u16` -> `u32` -> `u16` as before, but just `u16` all the way. That said, I do agree that this would warrant a minor version bump (or whatever `cargo-semver-checks` suggests :wink:).

fintelia commented:
Cargo semver-checks doesn't capture behavior changes. If there's downstream code that assumes (perhaps via `if let`) that it'll get a `Value::Unsigned()` for something and we switch the decoder to produce `Value::Short()` instead, the code will stop working.

weiji14 commented:
Ok, what would you recommend then to mitigate against downstream breakages? Would a minor version bump to v0.10.0 not suffice?

weiji14 committed:
Merge branch 'master' into short_u16

weiji14 committed:
Fix encode_decode tests returning Short instead of Unsigned for some tags

Some tags (Compression, PhotometricInterpretation, SamplesPerPixel, ResolutionUnit, Predictor) now return Short (u16) values instead of Unsigned (u32).

weiji14 commented on tests/encode_images.rs:
```diff
@@ -55,17 +55,17 @@ fn encode_decode() {
+                (Tag::Compression, ifd::Value::Short(1)),
+                (Tag::PhotometricInterpretation, ifd::Value::Short(2)),
                 (Tag::StripOffsets, ifd::Value::Unsigned(8)),
-                (Tag::SamplesPerPixel, ifd::Value::Unsigned(3)),
+                (Tag::SamplesPerPixel, ifd::Value::Short(3)),
                 (Tag::RowsPerStrip, ifd::Value::Unsigned(3334)),
                 (Tag::StripByteCounts, ifd::Value::Unsigned(30000)),
                 (Tag::XResolution, ifd::Value::Rational(1, 1)),
                 (Tag::YResolution, ifd::Value::Rational(1, 1)),
-                (Tag::ResolutionUnit, ifd::Value::Unsigned(1)),
+                (Tag::ResolutionUnit, ifd::Value::Short(1)),
                 (Tag::Artist, ifd::Value::Ascii("Image-tiff".into())),
-                (Tag::Predictor, ifd::Value::Unsigned(1)),
+                (Tag::Predictor, ifd::Value::Short(1)),
```
> The hard part is still [this concern](https://github.com/image-rs/image-tiff/issues/204#issuecomment-1566616975):
> 
> > This makes sense to fix, though the annoying part is figuring out whether anywhere else in the code is relying on the current behavior. Or whether it is likely to break any downstream users and thus might warrant additional caution

Looks like some tests added in #255 around November 2024 (after this PR was opened) discovered a potential area of breakage. Old behaviour is that some of these tags (e.g. Compression, PhotometricInterpretation, SamplesPerPixel, ResolutionUnit, Predictor) were incorrectly decoded as Unsigned (u32), whereas after this PR, they will be decoded as Short (u16).

weiji14 committed:
Raise TiffFormatError:ShortExpected in into_u16_vec

weiji14 committed:
More proper Short/SignedShort outputs for reading tags (case 4)

Adapted from https://github.com/developmentseed/async-tiff/pull/94/commits/23a6e5d7d273a2f5f24534e366b288c19e050d8d

fintelia reviewed (approved):
I don't think there's much more we can do to figure out potential breakage, so I'll go ahead and merge this unless there's any objections
